pilot/pkg/networking/plugin/plugin.go

```diff
@@ -47,7 +50,29 @@ const (
 )
 
 // ModelProtocolToListenerProtocol converts from a config.Protocol to its corresponding plugin.ListenerProtocol
-func ModelProtocolToListenerProtocol(protocol config.Protocol) ListenerProtocol {
+func ModelProtocolToListenerProtocol(node *model.Proxy, protocol config.Protocol) ListenerProtocol {
+	switch protocol {
+	case config.ProtocolHTTP, config.ProtocolHTTP2, config.ProtocolGRPC, config.ProtocolGRPCWeb:
+		return ListenerProtocolHTTP
+	case config.ProtocolTCP, config.ProtocolHTTPS, config.ProtocolTLS,
+		config.ProtocolMongo, config.ProtocolRedis, config.ProtocolMySQL:
+		return ListenerProtocolTCP
+	case config.ProtocolUnsupported:
+		if util.IsProxyVersionGE13(node) {
+			return ListenerProtocolAuto
+		}
+
+		return ListenerProtocolUnknown
+	default:
+		return ListenerProtocolUnknown
```

Comment:
This is still not right. The idea here is that unknown is never used or it’s an error. When a protocol is not http, it’s mostly protocolTCP. It’s here that we need to do the protocol detection.

Reply:
In order to support proxy with old versions, I think we should still use this one.

Reply:
When we confirm all proxies support protocol sniffing, it's safe to remove this.